src/library/dao/trackdao.cpp

```diff
@@ -2180,50 +2189,54 @@ void TrackDAO::detectCoverArtForTracksWithoutCover(volatile const bool* pCancel,
 TrackPointer TrackDAO::getOrAddTrack(
         const TrackRef& trackRef,
         bool* pAlreadyInLibrary) {
-    if (!trackRef.isValid()) {
-        return TrackPointer();
+    // This function should only be used when you do not yet know whether the
+    // specified track is contained in the library. Use TrackDAO::getTrackByRef
+    // instead for tracks that you already know to be in the library.
+    DEBUG_ASSERT(!trackRef.hasId());
+    VERIFY_OR_DEBUG_ASSERT(trackRef.hasLocation()) {
+        return {};
     }
-
-    const TrackId trackId = getTrackIdByRef(trackRef);
-    if (trackId.isValid()) {
-        const auto pTrack = getTrackById(trackId);
-        if (pTrack) {
+    TrackPointer pTrack = GlobalTrackCacheLocker().lookupTrackByRef(trackRef);
+    if (!pTrack) {
+        // track not cached
+        const TrackId trackId = getTrackIdByLocation(trackRef.getLocation());
+        if (trackId.isValid()) {
+            // create track from library
+            pTrack = getTrackById(trackId);
+        }
+    }
+    if (pTrack) {
+        if (pTrack->getId().isValid()) {
             DEBUG_ASSERT(pTrack->getDateAdded().isValid());
             if (pAlreadyInLibrary) {
                 *pAlreadyInLibrary = true;
             }
             return pTrack;
         }
-        if (!trackRef.hasLocation()) {
-            qWarning()
-                    << "Failed to get track"
-                    << trackRef;
-            return TrackPointer();
-        }
     }
 
-    DEBUG_ASSERT(trackRef.hasLocation());
     addTracksPrepare();
-    const auto pTrack = addTracksAddFile(trackRef.getLocation(), true);
-    addTracksFinish(!pTrack);
-    if (!pTrack) {
-        qWarning()
-                << "Failed to add track"
-                << trackRef;
-        return TrackPointer();
+    bool unremove = true;
+    const TrackPointer pAddedTrack = addTracksAddFile(trackRef.getLocation(), unremove);
+    bool rollback = !pAddedTrack;
+    addTracksFinish(rollback);
+    if (!pAddedTrack) {
+        return pAddedTrack;
     }
     if (pAlreadyInLibrary) {
         *pAlreadyInLibrary = false;
     }
-    DEBUG_ASSERT(pTrack->getDateAdded().isValid());
+    DEBUG_ASSERT(pAddedTrack->getDateAdded().isValid());
+    DEBUG_ASSERT(!pTrack || pTrack == pAddedTrack);
 
     // If the track wasn't in the library already then it has not yet
     // been checked for cover art.
-    const auto future = guessTrackCoverInfoConcurrently(pTrack);
-    // Don't wait for the result and keep running in the background
-    Q_UNUSED(future)
-
-    return pTrack;
+    if (!pAddedTrack->getCoverInfo().hasImage()) {
+        const auto future = guessTrackCoverInfoConcurrently(pAddedTrack);
+        // Don't wait for the result and keep running in the background
+        Q_UNUSED(future)
```

Comment:
nit: Is there value in storing the future in an unsed variable droped in the same scope? It feels a bit bloated to me
```suggestion
        // Don't wait for the result and keep running in the background
        guessTrackCoverInfoConcurrently(pAddedTrack);
```

Reply:
`guessTrackCoverInfoConcurrently` is marked as `[[nodiscard]]` - which is usually the right call when returning futures, as discarding them is usually indicative of an oversight. So the `Q_UNUSED` is required.

I wouldn't consider this bloat, as I find the current code nice & self-documenting.